### PR TITLE
Fixes slow roundstarts

### PR DESCRIPTION
**Pull request by AffectedArc07**
## What Does This PR Do
Moves a DB call into async territory at roundstart to avoid the start of the round taking a bit

## Why It's Good For The Game
Roundstarts shouldnt lag for 6 seconds for no reason

## Changelog
:cl: AffectedArc07
fix: Roundstart is faster now
/:cl:

### CODE REVIEW

File: code/game/gamemodes/game_mode.dm [hunk 1]
```diff
@@ -91,6 +85,17 @@
 /datum/game_mode/process()
 	return 0
 
+// I wonder what this could do guessing by the name
+/datum/game_mode/proc/set_mode_in_db()
+	if(SSticker && SSticker.mode && SSdbcore.IsConnected())
```
Comment (dearmochi) [inline code/game/gamemodes/game_mode.dm:90]:
```suggestion
	if(SSticker?.mode && SSdbcore.IsConnected())
```

File: code/game/gamemodes/game_mode.dm [hunk 2]
```diff
@@ -91,6 +85,17 @@
...
+		var/datum/db_query/query_round_game_mode = SSdbcore.NewQuery("UPDATE [format_table_name("round")] SET game_mode=:gm WHERE id=:rid", list(
+			"gm" = SSticker.mode,
```
Comment (dearmochi) [inline code/game/gamemodes/game_mode.dm:92]: https://github.com/ParadiseSS13/Paradise/blob/9480211187438f8e7a54d3d87b838531efe9af2c/code/controllers/subsystem/ticker.dm#L22-L23

`SSticker.mode` is a datum. How does that work?

Additionally, any reason you get the information from the ticker instead of the mode (`src`) itself?

Comment (AffectedArc07) [reply code/game/gamemodes/game_mode.dm:92]: It takes the `name` var, but I will force specify it. As for why I do the ticker lookup, sometimes with forced secret modes things can break, so this is a surefire way that it grabs the right thing.

**Review by dearmochi — APPROVED** [review]

File: code/game/gamemodes/game_mode.dm [hunk 3]
```diff
@@ -91,6 +85,17 @@
 /datum/game_mode/process()
 	return 0
 
+// I wonder what this could do guessing by the name
```
Comment (farie82) [inline code/game/gamemodes/game_mode.dm:88]: useless comment here
```suggestion
```

File: code/game/gamemodes/game_mode.dm [hunk 4]
```diff
@@ -91,6 +85,17 @@
...
+/datum/game_mode/proc/set_mode_in_db()
+	if(SSticker?.mode && SSdbcore.IsConnected())
+		var/datum/db_query/query_round_game_mode = SSdbcore.NewQuery("UPDATE [format_table_name("round")] SET game_mode=:gm WHERE id=:rid", list(
+			"gm" = SSticker.mode.name,
+			"rid" = GLOB.round_id
+		))
+		// We dont do anything with output. Dont bother wrapping with if()
+		query_round_game_mode.warn_execute()
+		qdel(query_round_game_mode)
```
Comment (farie82) [inline code/game/gamemodes/game_mode.dm:97]:
```suggestion
	if(!SSticker?.mode || !SSdbcore.IsConnected())
		return
	var/datum/db_query/query_round_game_mode = SSdbcore.NewQuery("UPDATE [format_table_name("round")] SET game_mode=:gm WHERE id=:rid", list(
		"gm" = SSticker.mode.name,
		"rid" = GLOB.round_id
	))
	// We dont do anything with output. Dont bother wrapping with if()
	query_round_game_mode.warn_execute()
	qdel(query_round_game_mode)
```

**Review by farie82 — CHANGES_REQUESTED** [review]
Why does this DB call need to be made async btw? Seems like a simple 1 record update which would take no time to execute

**Comment by AffectedArc07** [comment]
> Why does this DB call need to be async btw? Seems like a simple 1 record update which would take no time to execute

I will take any performance I can, that and it takes a full 6 seconds due to it being at roundstart, and ill explain why.

Async queries are "send execute request then poll for completion", and the polling is done once every tick, with the backbone being inside an `UNTIL()` statement.
```dm
	if(async)
		var/job_id = rustg_sql_query_async(connection, sql, json_encode(arguments))
		in_progress = TRUE
		UNTIL((job_result_str = rustg_sql_check_query(job_id)) != RUSTG_JOB_NO_RESULTS_YET)
		in_progress = FALSE
```
Until equates to
```dm
#define UNTIL(X) while(!(X)) stoplag()
```

And `stoplag()` equates to
```dm
/proc/stoplag(initial_delay)
	if(!Master || !(Master.current_runlevel & RUNLEVELS_DEFAULT))
		sleep(world.tick_lag)
		return 1
	if(!initial_delay)
		initial_delay = world.tick_lag
	. = 0
	var/i = DS2TICKS(initial_delay)
	do
		. += CEILING(i*DELTA_CALC, 1)
		sleep(i*world.tick_lag*DELTA_CALC)
		i *= 2
	while(TICK_USAGE > min(TICK_LIMIT_TO_RUN, Master.current_ticklimit))
```
TLDR of `stoplag()` is to sleep until the next tick, since a constant `while() continue` would eat up all available CPU cycles. It repeats this cycle each tick until tick usage is below the limit. Because roundstart is a massive amount of activity happening in one tick, the tick usage is massive (400%!), and stoplag takes a while until it balances out between the amount of times it has slept, and the tick target. This causes a large delay on roundstart because of this. By moving the saving to its own function which is invoked asyncly, the DB can take as long as it wants to save, while roundstart continues as normal.
